fog/view/server/src/fog_view_router_request_handler.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use crate::{
+    error::{router_server_err_to_rpc_status, RouterServerError},
+    fog_view_shard_responses_processor,
+};
+
+use futures::{future::try_join_all, SinkExt, TryStreamExt};
+use grpcio::{ChannelBuilder, DuplexSink, RequestStream, RpcStatus, WriteFlags};
+use mc_attest_api::attest;
+use mc_common::{logger::Logger, ResponderId};
+use mc_fog_api::{
+    view::{
+        FogViewRouterRequest, FogViewRouterResponse, MultiViewStoreQueryRequest,
+        MultiViewStoreQueryResponse,
+    },
+    view_grpc::FogViewApiClient,
+};
+use mc_fog_uri::FogViewUri;
+use mc_fog_view_enclave_api::ViewEnclaveProxy;
+use mc_util_grpc::{rpc_invalid_arg_error, ConnectionUriGrpcioChannel};
+use std::{str::FromStr, sync::Arc};
+
+const RETRY_COUNT: usize = 3;
+
+/// Handles a series of requests sent by the Fog Router client.
+pub async fn handle_requests<E>(
+    shard_clients: Vec<Arc<FogViewApiClient>>,
+    enclave: E,
+    mut requests: RequestStream<FogViewRouterRequest>,
+    mut responses: DuplexSink<FogViewRouterResponse>,
+    logger: Logger,
+) -> Result<(), grpcio::Error>
+where
+    E: ViewEnclaveProxy,
+{
+    while let Some(request) = requests.try_next().await? {
+        let result = handle_request(
```

Comment:
I believe that this is acceptable because this method represents the streaming connection between client and router. The client wants to get responses back in the order that it sends requests, so this should be fine. I think that the utility of async here is that the async executor (in our case the gRPC async executor) can schedule other work while this `try_next` yields. Moreover, this set-up is how the Rust gRPC async example implementation does it.